main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -957,6 +957,21 @@ object Typer {
           resultEff = eff
         } yield (constrs, resultTyp, resultPur, resultEff)
 
+      case KindedAst.Expression.OnExit(exp1, exp2, loc) =>
```

Comment:
What I understand here is that we allow the first argument to have any effect. This effect will eventually be executed upon exiting something, but it doesn't show up in the effect system. This feels a little strange to me.

Reply:
I understand why it feels strange. *In practice* I think it'll work out because anything `Closeable` will have non-trivial (probably, but not necessarily, `IO`) effects within the region.

But I agree that the effect being "swallowed" here isn't idea. I guess the effect should be added to the effect of the `Scope` with which this `onClose` is associated? I'm not sure how to achieve that though?

Reply:
A thought:

The thing *calling* `onClose` will always be within a region. If we add the effect of the function passed to `onClose` to the `onClose` itself, it will end up in the region. At the "wrong" point (the point where `onClose` is called, not the point at which the region terminates) but perhaps that is OK?

Reply:
I see. I think the ideal would be to have an effect that carries the idea that there's no effect _here_, but it will occur later, like `Latent[IO]` or even `Latent[r, IO]` to associate a region. Then upon reaching the end of the region, the effect is translated into the proper type.

Of course, our system cannot handle that kind of complexity. 

For now, I think the safest approximation is to put the effect on the `onClose`.

Reply:
To chime in, but in a different way:

Its important that ON_EXIT is not pure because if it is then the optimizer can remove it.

Right now it can be pure if e1 and e2 are. Can they be? The answer is yes, because they are a lambda and a region. 

So we need to change the `resultPur` to be `Impure`. (Or if you want `r and Impure`).